File: bot.py
#!/usr/bin/env python3
"""The paper-trading bot: a scheduling layer over allocator.py.

allocator.py already knows how to spread a simulated bankroll across a scan's
pairs -- what it does not do is run itself. It is a command you type once per
scan you have reviewed. This module is what turns that into a loop: after every
scan the server finishes, `run_cycle` settles what is due, plans an allocation,
and books it, then records one point on an equity curve.

Everything here is deliberately thin. `plan_allocation`, `apply_orders` and
`settle` are the allocator's and stay the allocator's, so the sizing rules the
Bot tab obeys are the same ones `python allocator.py allocate` obeys. What this
module adds is the three things a running bot needs and a one-shot command does
not: persistence of its own on/off state, mark-to-market on open positions, and
writes that survive being interrupted.

A word on what this bot is. The scanner reports *candidates*, not confirmed
arbitrages, and README.md spends a section on why: a cross-venue hedge is only
risk-free if both markets resolve on identical terms, and on the seed labels 18
of 24 candidates did not. Reading both rule texts is the step that catches that,
and it is exactly the step an auto-trader skips. So the defaults here are the
cautious ones -- rules must be compared and must agree, and settlement assumes
the seed base rate of mismatches rather than the flattering zero. Both are
settings; see DEFAULT_CONFIG.
"""

from __future__ import annotations

import json
import os
import tempfile
from datetime import datetime, timezone

import allocator

HERE = os.path.dirname(os.path.abspath(__file__))
BOT_PATH = os.path.join(HERE, "bot.json")

# How many cycles of history to keep. One per scan; at a scan every 15 minutes
# through waking hours that is a couple of months of equity curve.
CYCLE_LOG_LIMIT = 500

# The allocator's config plus the three settings that only mean something to a
# thing that runs on its own.
DEFAULT_CONFIG = dict(
    allocator.CONFIG,

    # What fraction of settling positions turn out to have been mismatches --
    # two different questions that merely read alike, where the "hedge" was
    # really two directional bets and both can lose.
    #
    # 0.0 is the optimistic world where every candidate was genuinely the same
    # question. The seed labels put the real rate near 0.75 for unfiltered
    # candidates, and that is the default here on purpose: a bot that assumes
    # zero reports a healthy profit by defining away the single largest risk in
    # the strategy. Lower it if you are filtering candidates by hand.
    mismatch_loss_rate=0.75,

    # Seed for the settlement draw. None is a genuinely random draw; set an
    # integer to make a run reproducible.
    settle_seed=None,

    # Settle due positions at the start of each cycle.
    auto_settle=True,
)


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


# ---------------------------------------------------------------------------
# Persistence
#
# Both files are written atomically. The allocator writes paper.json straight
# over the old one (allocator.save_portfolio), which is fine for a command that
# runs and exits, but this process serves HTTP while it writes: a crash or a
# read landing mid-write would leave a truncated portfolio and lose the whole
# simulated history. Write to a temp file in the same directory, then rename --
# os.replace is atomic, so a reader sees either the old file or the new one.
# ---------------------------------------------------------------------------

def write_json(path: str, payload) -> None:
    directory = os.path.dirname(os.path.abspath(path)) or "."
    os.makedirs(directory, exist_ok=True)
    handle, tmp = tempfile.mkstemp(dir=directory, prefix=".tmp-", suffix=".json")
    try:
        with os.fdopen(handle, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2)
        os.replace(tmp, path)
    except BaseException:
        # Leave the existing file untouched rather than half-written.
        try:
            os.remove(tmp)
        except OSError:
            pass
        raise


def default_state() -> dict:
    return {
        "running": False,
        "config": dict(DEFAULT_CONFIG),
        "last_cycle": None,
        "cycle_log": [],
    }


def load_state() -> dict:
    """The bot's own state. A missing or corrupt file is not fatal.

    A bot that refuses to start because its settings file was truncated is
    worse than one that starts from defaults and says so -- the portfolio, the
    part that actually matters, lives in paper.json and is loaded separately.
    """
    state = default_state()
    try:
        with open(BOT_PATH, encoding="utf-8") as fh:
            stored = json.load(fh)
    except (OSError, json.JSONDecodeError):
        return state

    if not isinstance(stored, dict):
        return state
    state["running"] = bool(stored.get("running", False))
    state["last_cycle"] = stored.get("last_cycle")
    log = stored.get("cycle_log")
    state["cycle_log"] = log[-CYCLE_LOG_LIMIT:] if isinstance(log, list) else []
    # Merge rather than replace, so a config written by an older version picks
    # up new keys with their defaults instead of raising a KeyError deep in the
    # allocator.
    if isinstance(stored.get("config"), dict):
        state["config"].update(stored["config"])
    return state


def save_state(state: dict) -> None:
    state["cycle_log"] = state["cycle_log"][-CYCLE_LOG_LIMIT:]
    write_json(BOT_PATH, state)


def load_portfolio(state: dict) -> dict:
    return allocator.load_portfolio(state["config"]["bankroll"])


def save_portfolio(portfolio: dict) -> None:
    write_json(allocator.PAPER_PATH, portfolio)


def clean_config(incoming: dict) -> dict:
    """Keep only known keys, coerced to the type of the default.

    This is the boundary where browser JSON becomes numbers the allocator
    divides by. An unvalidated string here surfaces as a TypeError halfway
    through sizing a position, with the portfolio already half-mutated.
    """
    cleaned = {}
    for key, value in (incoming or {}).items():
        if key not in DEFAULT_CONFIG:
            continue
        default = DEFAULT_CONFIG[key]
        try:
            if key == "settle_seed":
                cleaned[key] = None if value in (None, "") else int(value)
            elif isinstance(default, bool):
                cleaned[key] = bool(value)
            elif isinstance(default, float):
                cleaned[key] = float(value)
            elif isinstance(default, int):
                cleaned[key] = int(value)
            else:
                cleaned[key] = str(value)
        except (TypeError, ValueError):
            continue  # unparseable: keep the existing setting
    return cleaned


# ---------------------------------------------------------------------------
# Mark to market
# ---------------------------------------------------------------------------

def mark_to_market(portfolio: dict, records: list[dict]) -> list[dict]:
    """Open positions re-priced against the newest scan.

    The allocator values open positions at cost (see its `equity`), because a
    paper position is held to settlement and its true value is not known until
    it resolves. That is the right basis for sizing, but it makes a monitor
    useless: every open row would read zero forever.

    So each open position is matched back to its pair in the current scan by
    key, and valued at what the same hedge costs now. A pair that has dropped
    out of the scan -- expired, or fell below a threshold -- gets no mark at
    all rather than a stale one, and says so.
    """
    current = {allocator.pair_key(rec): rec for rec in records or []}
    marked = []
    for pos in portfolio["open"]:
        row = dict(pos)
        pair = current.get(pos["key"])
        cost_now = (pair or {}).get("cost_per_contract")
        if pair is None or cost_now is None or not (0.0 < cost_now < 1.0):
            row["mark"] = None
            row["unrealized"] = None
            row["mark_note"] = "no current quote"
        else:
            value = round(pos["contracts"] * cost_now, 4)
            row["mark"] = value
            row["unrealized"] = round(value - pos["stake"], 4)
            row["mark_note"] = ""
            row["edge_now"] = pair.get("edge")
        marked.append(row)
    return marked


def unrealized_pnl(marked: list[dict]) -> float:
    return round(sum(row["unrealized"] for row in marked
                     if row.get("unrealized") is not None), 4)


# ---------------------------------------------------------------------------
# The cycle
# ---------------------------------------------------------------------------

def run_cycle(records: list[dict], state: dict, portfolio: dict) -> dict:
    """One bot cycle against a fresh scan. Mutates and saves the portfolio.

    Settle first, then allocate: a position resolving today frees cash that the
    same cycle can put back to work, which is what a person doing this by hand
    would do.
    """
    cfg = state["config"]
    settled = []
    if cfg["auto_settle"]:
        settled = allocator.settle(portfolio, cfg["mismatch_loss_rate"],
                                   cfg["settle_seed"])

    orders, notes = allocator.plan_allocation(portfolio, records, cfg)
    placed = allocator.apply_orders(portfolio, orders)
    if orders and not placed:
        notes.append("orders planned but none fit the remaining cash")

    save_portfolio(portfolio)

    cycle = summarise(portfolio, records, opened=len(placed),
                      settled=len(settled), notes=notes)
    state["last_cycle"] = cycle
    state["cycle_log"].append(cycle)
    save_state(state)
    return cycle


def summarise(portfolio: dict, records: list[dict], opened: int = 0,
              settled: int = 0, notes: list[str] | None = None) -> dict:
    """One point on the equity curve, plus what happened to get there."""
    marked = mark_to_market(portfolio, records)
    return {
        "at": now_iso(),
        "equity": round(allocator.equity(portfolio), 2),
        "cash": round(portfolio["cash"], 2),
        "exposure": round(allocator.open_stake(portfolio), 2),
        "realized": allocator.realized_pnl(portfolio),
        "unrealized": unrealized_pnl(marked),
        "open": len(portfolio["open"]),
        "closed": len(portfolio["closed"]),
        "opened": opened,
        "settled": settled,
        "notes": notes or [],
    }


# ---------------------------------------------------------------------------
# Manual actions
# ---------------------------------------------------------------------------

def flatten(portfolio: dict) -> list[dict]:
    """Close every open position at cost. Nothing won, nothing lost.

    Distinct from `settle`, which resolves positions whose date has passed and
    pays out or wipes out accordingly. This is the "stop, I want my simulated
    cash back" button: it makes no claim about how those markets would have
    resolved, so it returns the stake and books zero P&L.
    """
    closed = []
    for pos in portfolio["open"]:
        portfolio["cash"] = round(portfolio["cash"] + pos["stake"], 4)
        closed.append(dict(pos, status="flattened", payoff=pos["stake"],
                           pnl=0.0, settled=now_iso()))
    portfolio["open"] = []
    portfolio["closed"].extend(closed)
    save_portfolio(portfolio)
    return closed


def reset(bankroll: float) -> dict:
    """Throw the portfolio away and start over at `bankroll`."""
    portfolio = {
        "starting_bankroll": bankroll,
        "cash": bankroll,
        "open": [],
        "closed": [],
        "created": now_iso(),
    }
    save_portfolio(portfolio)
    return portfolio


def snapshot(state: dict, portfolio: dict, records: list[dict]) -> dict:
    """Everything the Bot tab draws, in one payload."""
    marked = mark_to_market(portfolio, records)
    return {
        "running": state["running"],
        "config": state["config"],
        "starting_bankroll": portfolio["starting_bankroll"],
        "equity": round(allocator.equity(portfolio), 2),
        "cash": round(portfolio["cash"], 2),
        "exposure": round(allocator.open_stake(portfolio), 2),
        "realized": allocator.realized_pnl(portfolio),
        "unrealized": unrealized_pnl(marked),
        "open": marked,
        "closed": portfolio["closed"][-100:],
        "closed_total": len(portfolio["closed"]),
        "last_cycle": state["last_cycle"],
        "cycle_log": state["cycle_log"],
        "created": portfolio.get("created"),
    }

File: dashboard/bot.js
/* ===========================================================================
   The Bot tab — the paper-trading loop, watched.

   Only reachable in live mode: the numbers here come from /api/bot, which is
   allocator.py's portfolio plus the bot's own state. Nothing on this page
   places a real order, holds a real position, or needs an account.

   The honest-reporting rule for this view: a simulated bankroll that only ever
   goes up is a bug in the simulation, not a strategy. The scanner reports
   *candidates*, and on the seed labels three quarters of them were different
   questions that merely read alike — so the mismatch rate is a first-class
   control here rather than a hidden assumption, and the panel says what it
   costs. See the "Why it is paper only" section of the README.
   =========================================================================== */

(function (App) {
  "use strict";

  const $ = (id) => document.getElementById(id);
  const el = (tag, cls, text) => {
    const node = document.createElement(tag);
    if (cls) node.className = cls;
    if (text !== undefined) node.textContent = text;
    return node;
  };
  const money = (v) => (v === null || v === undefined) ? "—" : "$" + Number(v).toFixed(2);
  const signed = (v) => (v === null || v === undefined) ? "—"
    : (v > 0 ? "+" : v < 0 ? "−" : "") + "$" + Math.abs(Number(v)).toFixed(2);
  const pct = (v, d = 1) => (v === null || v === undefined) ? "—" : (v * 100).toFixed(d) + "%";
  const tone = (v) => (v === null || v === undefined || Math.abs(v) < 0.005) ? ""
    : (v > 0 ? "pos" : "neg");
  const shortTime = (iso) => (iso || "").replace("T", " ").slice(0, 16);

  // key, label, min, max, step, hint. The order money flows through them.
  const KNOBS = [
    ["bankroll", "Starting bankroll", 100, 100000, 100,
     "Simulated cash. Takes effect on reset."],
    ["positions", "Positions held at once", 1, 30, 1,
     "The spread. More positions, less riding on any one being a real match."],
    ["deploy_frac", "Deployed per cycle", 0, 1, 0.05,
     "Share of equity put to work each scan; the rest stays in reserve."],
    ["max_position_frac", "Cap per position", 0.01, 1, 0.01,
     "Hard ceiling on any single pair, as a fraction of equity."],
    ["time_pref", "Time preference", 0, 3, 0.1,
     "Above 0 prejudices sooner-resolving pairs: edge ÷ days^pref. 0 sizes on edge alone."],
    ["min_edge", "Minimum edge", 0, 0.10, 0.005,
     "Locked profit per $1 after fees, below which a pair is not worth the capital."],
    ["min_confidence", "Minimum confidence", 0, 1, 0.05,
     "Text-match floor. Deliberately stricter than the board's — this is money, pretend or not."],
    ["mismatch_loss_rate", "Assumed mismatch rate", 0, 1, 0.05,
     "The share of settling positions that turn out to be two different questions, " +
     "where the hedge was really two bets and the stake is lost. The seed labels put " +
     "this near 0.75 for candidates nobody read by hand. At 0 the same eight pairs " +
     "return +11%; at 0.75, −58% to −99%. This is the single biggest number on the page."],
  ];

  /* ── header ──────────────────────────────────────────────────────────── */

  function header(snap) {
    const grid = el("div", "stat-grid");
    const start = snap.starting_bankroll || 0;
    const total = snap.equity + snap.unrealized;
    const ret = start ? (total - start) / start : null;

    grid.append(
      App.statTile("equity", money(snap.equity)),
      App.statTile("cash", money(snap.cash)),
      App.statTile("exposure", money(snap.exposure)),
      App.statTile("realized", signed(snap.realized), tone(snap.realized)),
      App.statTile("unrealized", signed(snap.unrealized), tone(snap.unrealized)),
      App.statTile("return", ret === null ? "—" : (ret > 0 ? "+" : "") + pct(ret),
                   tone(ret)),
      App.statTile("open", String(snap.open.length)),
      App.statTile("closed", String(snap.closed_total)));
    return grid;
  }

  function controls(snap) {
    const bar = el("div", "botbar");

    const state = el("span", "botstate " + (snap.running ? "on" : "off"));
    state.append(el("span", "dot " + (snap.running ? "fresh" : "stale")));
    state.append(el("span", undefined, snap.running ? "running" : "paused"));
    bar.append(state);

    bar.append(el("span", "muted",
      snap.running
        ? "Trades every scan that finishes."
        : "Scans still run; the bot just does not act on them."));

    const spacer = el("div", "toolbar-spacer");
    bar.append(spacer);

    const act = (label, action, cls, confirmText) => {
      const button = el("button", "btn " + (cls || ""), label);
      button.addEventListener("click", async () => {
        if (confirmText && !confirm(confirmText)) return;
        button.disabled = true;
        try {
          await App.post("/api/bot", { action });
          await App.reloadState();
          App.renderBot();
        } catch (err) {
          note("Could not " + label.toLowerCase() + ": " + err.message);
          button.disabled = false;
        }
      });
      return button;
    };

    bar.append(snap.running ? act("Pause", "pause", "ghost")
                            : act("Start", "start", "primary"));
    bar.append(act("Trade this scan", "allocate", "ghost"));
    bar.append(act("Settle what's due", "settle", "ghost"));
    bar.append(act("Flatten", "flatten", "ghost",
      "Close all " + snap.open.length + " open positions at cost, booking no profit or loss?"));
    bar.append(act("Reset", "reset", "danger",
      "Throw away the portfolio and its whole history, and start again at "
      + money(snap.config.bankroll) + "?"));
    return bar;
  }

  function note(text) {
    const banner = $("banner-error");
    banner.hidden = false;
    banner.textContent = text;
  }

  /* ── open positions ──────────────────────────────────────────────────── */

  function positions(snap) {
    if (!snap.open.length) {
      return App.panel("Open positions",
        el("p", "panel-note",
           snap.running
             ? "Nothing held. The next scan that turns up an eligible pair will open one."
             : "Nothing held, and the bot is paused."));
    }

    const table = el("table", "plain");
    const head = el("tr");
    for (const [label, cls] of [["Opened", ""], ["Pair", ""], ["Leg", ""],
                                ["Cost", "num"], ["Qty", "num"], ["Stake", "num"],
                                ["Mark", "num"], ["Unreal.", "num"],
                                ["Edge", "num"], ["Settles", "num"]]) {
      head.append(el("th", cls, label));
    }
    const thead = el("thead");
    thead.append(head);
    table.append(thead);

    const body = el("tbody");
    for (const pos of snap.open) {
      const row = el("tr");
      row.append(el("td", "muted", shortTime(pos.opened)));

      const pair = el("td");
      // Position side fields follow the pair's generic a/b; fall back to the
      // old kalshi/poly names so a position opened before the rename still shows.
      pair.append(el("div", undefined, pos.a_q || pos.a_id || pos.kalshi_q || pos.kalshi_id));
      pair.append(el("div", "faint", pos.b_q || pos.b_id || pos.poly_q || pos.poly_url));
      row.append(pair);

      row.append(el("td", "faint", pos.leg || "—"));
      row.append(el("td", "num", Number(pos.cost_per_contract).toFixed(3)));
      row.append(el("td", "num", String(pos.contracts)));
      row.append(el("td", "num", money(pos.stake)));
      row.append(el("td", "num" + (pos.mark === null ? " faint" : ""),
                    pos.mark === null ? "no quote" : money(pos.mark)));
      row.append(el("td", "num " + tone(pos.unrealized),
                    pos.unrealized === null ? "—" : signed(pos.unrealized)));
      row.append(el("td", "num", pos.edge === null || pos.edge === undefined
                    ? "—" : (pos.edge * 100).toFixed(2)));
      row.append(el("td", "num muted", shortTime(pos.settle_time) || "—"));
      body.append(row);
    }
    table.append(body);

    const wrap = el("div", "tablewrap");
    wrap.append(table);
    return App.panel("Open positions", wrap, el("p", "panel-note",
      "Mark is what the same hedge costs in the latest scan — what unwinding now " +
      "would return. A pair that has dropped out of the scan gets no mark rather " +
      "than a stale one."));
  }

  /* ── equity curve ────────────────────────────────────────────────────── */

  function curve(snap) {
    const log = snap.cycle_log || [];
    if (log.length < 2) {
      return App.panel("Equity",
        el("p", "panel-note",
           "Two cycles are needed before there is a line to draw. "
           + (log.length ? "One so far." : "None so far.")));
    }
    const points = log.map((c) => ({
      label: shortTime(c.at),
      value: c.equity + (c.unrealized || 0),
    }));
    return App.panel("Equity", App.lineChart(points, (v) => "$" + Number(v).toFixed(0)),
      el("p", "panel-note", "Cash plus cost basis plus the unrealized mark, one point per cycle."));
  }

  /* ── closed trades ───────────────────────────────────────────────────── */

  function tradeLog(snap) {
    if (!snap.closed.length) {
      return App.panel("Trade log", el("p", "panel-note", "Nothing has settled yet."));
    }
    const table = el("table", "plain");
    const head = el("tr");
    for (const [label, cls] of [["Settled", ""], ["Pair", ""], ["Outcome", ""],
                                ["Stake", "num"], ["Payoff", "num"], ["P&L", "num"]]) {
      head.append(el("th", cls, label));
    }
    const thead = el("thead");
    thead.append(head);
    table.append(thead);

    const body = el("tbody");
    for (const pos of snap.closed.slice().reverse()) {
      const row = el("tr");
      row.append(el("td", "muted", shortTime(pos.settled)));
      row.append(el("td", undefined, pos.a_q || pos.a_id || pos.kalshi_q || pos.kalshi_id));

      const label = { "hedged-win": "hedge held", "mismatch-loss": "mismatch",
                      "flattened": "flattened" }[pos.status] || pos.status;
      const cls = pos.status === "mismatch-loss" ? "flag warn"
                : pos.status === "hedged-win" ? "flag" : "flag";
      const cell = el("td");
      cell.append(el("span", cls, label));
      row.append(cell);

      row.append(el("td", "num", money(pos.stake)));
      row.append(el("td", "num", money(pos.payoff)));
      row.append(el("td", "num " + tone(pos.pnl), signed(pos.pnl)));
      body.append(row);
    }
    table.append(body);

    const wrap = el("div", "tablewrap");
    wrap.append(table);
    const shown = snap.closed.length < snap.closed_total
      ? `Showing the last ${snap.closed.length} of ${snap.closed_total}.` : "";
    return App.panel("Trade log", wrap, el("p", "panel-note", shown));
  }

  /* ── rules ───────────────────────────────────────────────────────────── */

  function rules(snap) {
    const cfg = snap.config;
    const box = el("div");

    for (const [key, label, min, max, step, hint] of KNOBS) {
      const slider = el("div", "slider");
      const head = el("div", "slider-head");
      const shown = key === "bankroll" ? money(cfg[key])
                  : key === "positions" ? String(cfg[key])
                  : Number(cfg[key]).toFixed(key === "min_edge" ? 3 : 2);
      const value = el("span", "slider-val"
        + (key === "mismatch_loss_rate" && cfg[key] < 0.2 ? " changed" : ""), shown);
      head.append(el("span", "slider-name", label), value);

      const input = el("input");
      input.type = "range";
      input.min = min; input.max = max; input.step = step;
      input.value = cfg[key];
      // Redraw the number as it drags, but only tell the server on release —
      // otherwise a single drag is fifty POSTs and fifty portfolio writes.
      input.addEventListener("input", () => {
        cfg[key] = Number(input.value);
        value.textContent = key === "bankroll" ? money(cfg[key])
                          : key === "positions" ? String(cfg[key])
                          : Number(cfg[key]).toFixed(key === "min_edge" ? 3 : 2);
      });
      input.addEventListener("change", () => save({ [key]: Number(input.value) }));

      slider.append(head, input, el("div", "slider-hint", hint));
      box.append(slider);
    }

    box.append(toggle("Require the rule texts to agree", "require_rules_match", cfg,
      "The one check that separates a hedge from two directional bets. Off, the bot " +
      "will trade pairs whose resolution terms were never compared."));
    box.append(toggle("Only depth-verified pairs", "skip_top_of_book_only", cfg,
      "Skip pairs whose size was estimated from top-of-book instead of a real order book."));
    box.append(toggle("Settle due positions automatically", "auto_settle", cfg,
      "Resolve positions whose date has passed at the start of each cycle."));

    return App.panel("Rules", box, el("p", "panel-note",
      "The same sizing rules `python allocator.py allocate` obeys — this panel just " +
      "sets them from the browser."));
  }

  function toggle(label, key, cfg, hint) {
    const wrap = el("label", "toggle");
    const input = el("input");
    input.type = "checkbox";
    input.checked = Boolean(cfg[key]);
    input.addEventListener("change", () => {
      cfg[key] = input.checked;
      save({ [key]: input.checked });
    });
    wrap.append(input, el("span", "toggle-label", label));
    const box = el("div", "slider");
    box.append(wrap, el("div", "slider-hint", hint));
    return box;
  }

  async function save(patch) {
    try {
      await App.post("/api/bot", { action: "config", config: patch });
      await App.reloadState();
    } catch (err) {
      note("Could not save that setting: " + err.message);
    }
  }

  /* ── last cycle ──────────────────────────────────────────────────────── */

  function lastCycle(snap) {
    const cycle = snap.last_cycle;
    if (!cycle) {
      return App.panel("Last cycle", el("p", "panel-note",
        "The bot has not run a cycle yet."));
    }
    const rows = el("div", "waterfall");
    const line = (label, value, cls) => {
      const row = el("div", "wf-row" + (cls ? " " + cls : ""));
      row.append(el("span", undefined, label), el("span", "num", value));
      rows.append(row);
    };
    line("ran", shortTime(cycle.at));
    line("positions opened", String(cycle.opened));
    line("positions settled", String(cycle.settled));
    line("equity", money(cycle.equity), "total");

    const children = [rows];
    if ((cycle.notes || []).length) {
      const list = el("ul", "notes");
      for (const text of cycle.notes) list.append(el("li", undefined, text));
      children.push(list);
    }
    return App.panel("Last cycle", ...children);
  }

  /* ── the run log ─────────────────────────────────────────────────────── */

  function runLog() {
    const lines = (App.run && App.run.log) || [];
    if (!lines.length) return null;
    const pre = el("pre", "runlog", lines.slice(-60).join("\n"));
    return App.panel("Scanner output", pre);
  }

  /* ── the view ────────────────────────────────────────────────────────── */

  App.renderBot = function renderBot() {
    const view = $("view-bot");
    view.replaceChildren();

    const snap = App.bot;
    if (!snap) {
      view.append(App.panel("Bot", el("p", "panel-note",
        "No backend. Run `python app.py` to trade the scans as they land.")));
      return;
    }

    view.append(controls(snap));
    view.append(header(snap));

    const banner = el("p", "panel-note bot-caveat");
    banner.textContent =
      "Simulated money. No account, no wallet, no order ever leaves this machine. "
      + "Settlement assumes " + pct(snap.config.mismatch_loss_rate, 0)
      + " of positions turn out to be mismatched questions"
      + (snap.config.require_rules_match ? ", and pairs whose rule texts disagree are refused."
                                         : " — and rule-text agreement is NOT being required.");
    view.append(banner);

    view.append(positions(snap));

    const grid = el("div", "grid2");
    grid.append(curve(snap), lastCycle(snap));
    view.append(grid);

    view.append(tradeLog(snap));
    view.append(rules(snap));

    const log = runLog();
    if (log) view.append(log);

    const count = $("tab-count-bot");
    if (count) count.textContent = snap.open.length ? String(snap.open.length) : "";
  };

}(window.App));
